File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int[] map = new int[256]; // ASCII characters
        
        // initialize with -1
        for(int i = 0; i < 256; i++){
            map[i] = -1;
        }
        
        int left = 0;
        int maxLen = 0;
        
        for(int right = 0; right < s.length(); right++) {
            
            // if duplicate found
            if(map[s.charAt(right)] >= left) {
                left = map[s.charAt(right)] + 1;
            }
            
            // store index
            map[s.charAt(right)] = right;
            
            // update max length
            maxLen = Math.max(maxLen, right - left + 1);
        }
        
        return maxLen;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
            unordered_set<char> st;
        
        int left = 0;
        int right = 0;
        int maxLen = 0;
        
        while(right < s.length()) {
            
            // if duplicate found, shrink window
            while(st.find(s[right]) != st.end()) {
                st.erase(s[left]);
                left++;
            }
            
            // insert current character
            st.insert(s[right]);
            
            // update max length
            maxLen = max(maxLen, right - left + 1);
            
            right++;
        }
        
        return maxLen;   
    }
};